td-shim/src/acpi.rs

```diff
@@ -104,13 +104,19 @@ impl Xsdt {
     }
 
     pub fn add_table(&mut self, addr: u64) {
-        let table_num =
-            (self.header.length as usize - size_of::<GenericSdtHeader>()) / size_of::<u64>();
-        if table_num < ACPI_TABLES_MAX_NUM {
-            self.tables[table_num] = addr as u64;
-            self.header.length += size_of::<u64>() as u32;
+        if self.header.length < size_of::<GenericSdtHeader>() as u32 {
+            log::error!(
+                "Invalid header: Xsdt header length should not be less than generic header size"
+            );
         } else {
-            log::error!("too many ACPI tables, max {}", ACPI_TABLES_MAX_NUM);
+            let table_num =
+                (self.header.length as usize - size_of::<GenericSdtHeader>()) / size_of::<u64>();
+            if table_num < ACPI_TABLES_MAX_NUM {
+                self.tables[table_num] = addr as u64;
+                self.header.length += size_of::<u64>() as u32;
+            } else {
+                log::error!("too many ACPI tables, max {}", ACPI_TABLES_MAX_NUM);
```

Comment:
This is hard to unit test so I do wonder if either `add_table()` should return a `Result<>`, or if we need something like:

```suggestion
                #[cfg(not(test))]
                log::error!("too many ACPI tables, max {}", ACPI_TABLES_MAX_NUM);

                #[cfg(test)]
                panic!("too many ACPI tables, max {}", ACPI_TABLES_MAX_NUM);
```

You could then use a `#[should_panic(...)]` on a unit test to assert the expected behaviour.

Reply:
I added test case to check the table value with different add_table() scenarios.